### PR TITLE
Make a API calls and return the response

### PR DESCRIPTION
**Pull request by sulthan-ahmed**
Using two apis axos and also the SmartSurveyClient.  

- Added the tests

### CODE REVIEW

File: lib/api-call.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+const axios = require('axios');
+
+module.exports = {
+  getResponses: function(baseUrl, surveyID, endPoint, apiToken, apiTokenSecret) {
+    axios.get(baseUrl + '/' + surveyID + '/' + endPoint + '?api_token=' + apiToken + '&api_token_secret=' + apiTokenSecret)
```
Comment (sulthan-ahmed) [inline lib/api-call.js]: I would prefer this to be done in es6 format `${baseURL}/${surveyID}...` but it complained when attempting to do so

I would be grateful for ideas to make this look prettier

File: app.js [hunk 2]
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
 const hoganExpressStrict = require('hogan-express-strict');
 const expressPartialTemplates = require('express-partial-templates');
 const _ = require('lodash');
+const apiCall = require('./lib/api-call');
+const config = require('./config');
+const baseUrl = config.apiBaseUrl;
+const surveyID = config.surveyID;
+const responses = config.endPointResponses;
+const apiToken = config.apiToken;
+const apiTokenSecret = config.apiTokenSecret;
```
Comment (sulthan-ahmed) [inline app.js]: I had the above const in the api-call file itself but I found it tricky to tests so I've moved them

File: app.js [hunk 3]
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
 const hoganExpressStrict = require('hogan-express-strict');
 const expressPartialTemplates = require('express-partial-templates');
 const _ = require('lodash');
+const apiCall = require('./lib/api-call');
```
Comment (chrisns) [inline app.js]: `smartSurveyAPICall` or similar would be a better name, you might have other apis

File: app.js [hunk 4]
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
...
+const config = require('./config');
+const baseUrl = config.apiBaseUrl;
```
Comment (chrisns) [inline app.js]: if you *have* to do this rather than refer to `config.var` then consider object destructuring

File: config.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+/* eslint no-process-env: 0 */
+module.exports = {
+  apiToken: process.env.API_TOKEN,
+  apiTokenSecret: process.env.API_TOKEN_SECRET,
+  apiBaseUrl: process.env.BASE_URL || 'https://api.smartsurvey.io/v1/surveys',
+  surveyID: process.env.SURVEY_ID || '309701',
```
Comment (chrisns) [inline config.js]: you probably don't want to hardcode that as a default survey to use

File: lib/api-call.js (hunk 1 above)
Comment (chrisns) [reply lib/api-call.js]: you could do with something to build the GET url query params up based on an object.
https://nodejs.org/api/querystring.html
`querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' })`

for the building the `baseUrl + '/' + surveyID + '/' + endPoint` to check you're doing: 
```
`${baseUrl}/${surveyID}/${endPoint}`
```
yeh?

File: lib/api-call.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+'use strict';
```
Comment (chrisns) [inline lib/api-call.js]: consider making this into a separate node package, or at least keeping it decoupled (as you have done) from the rendering side of things so that you can do that.

**Review by chrisns — CHANGES_REQUESTED** [review]
did you look at https://www.npmjs.com/package/smartsurvey-client is it no good?

File: app.js (hunk 2 above)
Comment (daniel-ac-martin) [reply app.js]: Good idea to move them. I wouldn't bother setting consts here though. Unless you are using them multiple times.

Personally, I'd just do:
```js
apiCall.getResponses(
  config.baseUrl,
  config.surveyID,
  config.responses,
  config.apiToken,
  config.apiTokenSecret);
```

File: lib/api-call.js (hunk 1 above)
Comment (daniel-ac-martin) [reply lib/api-call.js]: @chrisns's solution is also nice because it will fix a minor bug in your code. Specifically, anything that goes into a URL should first be escaped / URL encoded.

File: lib/api-call.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+const axios = require('axios');
+
+module.exports = {
+  getResponses: function(baseUrl, surveyID, endPoint, apiToken, apiTokenSecret) {
+    axios.get(baseUrl + '/' + surveyID + '/' + endPoint + '?api_token=' + apiToken + '&api_token_secret=' + apiTokenSecret)
+      .then(function (response) {
+        console.log(response);
+      })
+      .catch(function (error) {
+        console.log(error);
```
Comment (daniel-ac-martin) [inline lib/api-call.js]: Would be nice to have more sophisticated logging that just console.log but perhaps it probably doesn't matter for a first iteration.

**Review by daniel-ac-martin — CHANGES_REQUESTED** [review]
Minor potential URL encoding bug.

**Comment by daniel-ac-martin** [comment]
I'd also suggest that the tests go in with this PR as it sounds like they pertain to this code. They shouldn't really go in as a separate PR in my opinion.

File: lib/api-call.js (hunk 6 above)
Comment (easternbloc) [reply lib/api-call.js]: also the file name. be more descriptive.

File: lib/api-call.js (hunk 1 above)
Comment (sulthan-ahmed) [reply lib/api-call.js]: good idea @chrisns 
yeah it was backtick but my sublime editor was complaining about es6 syntax.  

👍 @daniel-ac-martin

File: app.js (hunk 2 above)
Comment (sulthan-ahmed) [reply app.js]: nice one

File: app.js (hunk 3 above)
Comment (sulthan-ahmed) [reply app.js]: ahh good spot @chrisns about the npm package, I should've checked for that, I'll give it a try

File: app.js (hunk 4 above)
Comment (sulthan-ahmed) [reply app.js]: config.var :thinking: sorry basic question time, I've googled config.var and err I didn't quite find anything except referring to variables in my config so didn't quite understand this 

object deconstructing? ahh just read up about it, I did not know that quality, you mean this right? http://wesbos.com/destructuring-objects/

File: config.js (hunk 5 above)
Comment (sulthan-ahmed) [reply config.js]: this was pointing to the test version of the survey data so it made my life easier to have it default to that, are you saying I shouldn't have a default like that? 
The plan was to default to the live one

File: app.js (hunk 4 above)
Comment (chrisns) [reply app.js]: sorry I meant more `config.*` like @daniel-ac-martin 's comment about
```js
apiCall.getResponses(
  config.baseUrl,
....
```

File: app.js (hunk 3 above)
Comment (sulthan-ahmed) [reply app.js]: fair point about the name

File: app.js (hunk 4 above)
Comment (chrisns) [reply app.js]: yes, thats what I meant about destructuring :)

Comment (sulthan-ahmed) [reply app.js]: 👍 is this also object deconstructing or it doesn't have a special name?

**Comment by sulthan-ahmed** [comment]
@daniel-ac-martin I get what you're saying about including my test in this PR but I wanted to get thoughts about this PR while I work on the test.   That's already kind of proved quite handy

File: app.js (hunk 4 above)
Comment (chrisns) [reply app.js]: @daniel-ac-martin 's suggestion I don't think has a special name(?) its definitely not destructuring

**Comment by chrisns** [comment]
TDD ftw :)

**Comment by sulthan-ahmed** [comment]
Fair, next time I'll do a WIP

File: lib/smartSurveyAPICall.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  parseResponse: (err,result) => {
+    if (err) {
+      console.log(err);
+    }
+    else {
+      const prettyResult = JSON.stringify(result, null, 2);
+      console.log(prettyResult);
+    }
+  }
+};
```
Comment (sulthan-ahmed) [inline lib/smartSurveyAPICall.js]: You might be wondering why I am using two different apis.

I am using both as I'm experimenting with the usage of both, axios is more flexible and the smartSurveyClient is missing some methods that I may need.

File: package.json [hunk 9]
```diff
@@ -5,7 +5,7 @@
   "main": "app.js",
   "scripts": {
     "start": "node .",
-    "test": "NODE_ENV=test mocha",
+    "test": "NODE_ENV=test mocha test/*",
```
Comment (sulthan-ahmed) [inline package.json]: strangely I have to tell mocha the location of the directory

File: test/lib/smartSurveyAPICall.test.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+const proxy = require('proxyquire');
+const sinon = require('sinon');  
+const chai = require('chai').use(require('sinon-chai'));
+const should = chai.should();
+const proxyquire = require('proxyquire');
+let stubAxios = {};
+
+// Proxyquire syntax: first argument: point to your module (which holds your dependency)
+// 2nd argument: An object with the key: module name to proxy & the value the empty object above
+// So this means smartSurveyAxiosAPI will now point to this empty object, i.e. a stub
```
Comment (sulthan-ahmed) [inline test/lib/smartSurveyAPICall.test.js]: These comments are intentional for learning

File: test/lib/smartSurveyAPICall.test.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+const smartSurveyAxiosAPI = proxyquire('../../lib/smartSurveyAPICall', {
+  'axios': stubAxios,
+});
+
+// create an empty function using echma 5 syntax because of using the this context
+var stubSmartSurvey = function(params) {
```
Comment (sulthan-ahmed) [inline test/lib/smartSurveyAPICall.test.js]: getting strange behaviour when I use echma 6 syntax for context stuff

File: test/lib/smartSurveyAPICall.test.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      it('calls this.client.getResponses with the correct params', function() {
+        that.client.getResponses.should.have.been.calledWithExactly("1",{ page: 1, pageSize: 25, includeLabels: false}, that.parseResponse)
+      })
+      it('sets this.client to a new instance of SmartSurveyClient', function() {
+          let that = {};
+          // you can mock smartSurveyClientAPI as a constructor if you want to, but not essential
+          smartSurveyClientAPI.createClient.call(that, "a", "b");
+          that.client.should.be.instanceof(stubSmartSurvey);
+      })
+      it('calls the SmartSurveyClient constructor with the correct params', function() {
```
Comment (sulthan-ahmed) [inline test/lib/smartSurveyAPICall.test.js]: this test is failing
1. is it relevant?
2. how do I fix it?

File: test/lib/smartSurveyAPICall.test.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+          var stub = sinon.stub(stubSmartSurvey, 'constructor')
+
+          stubSmartSurvey.prototype.constructor = sinon.stub()
+
+          smartSurveyClientAPI.createClient.call(that, "1", "2");
+          stubSmartSurvey.constructor.should.have.been.calledWithExactly({ apiToken: '1', apiTokenSecret: '2' })
+      })
+
+      it('SmartSurveyAPItest: when there is an error, log it out', function() {
+        consoleSpy = sinon.stub(console, 'log');
```
Comment (sulthan-ahmed) [inline test/lib/smartSurveyAPICall.test.js]: this test is not running properly because of the stub on the console

File: package.json (hunk 9 above)
Comment (danielepolencic) [reply package.json]: *I think* this is due to the fact that you have a `lib` folder in your `test` folder. You could remove the `lib` folder or add `mocha --recursive`.

File: test/lib/smartSurveyAPICall.test.js (hunk 11 above)
Comment (danielepolencic) [reply test/lib/smartSurveyAPICall.test.js]: ECMA. Can you provide an example?

File: test/lib/smartSurveyAPICall.test.js (hunk 12 above)
Comment (danielepolencic) [reply test/lib/smartSurveyAPICall.test.js]: 1. There're few source of confusion in this test.
  - It's a good idea to test the invocation, but it's duplication to test that SmartSurveyClient behaves the way you want. Testing if the constructors receives the arguments correctly is a test that should be part of the library, not the code using it.
  - You stubbed the class and tested the invocation of the stubbed method. This test won't help you in the long run, because if the library changes the signature of the class or constructor, your test will still pass. You stubbed too much 😄 
2. I'd delete the test.

File: test/lib/smartSurveyAPICall.test.js (hunk 13 above)
Comment (danielepolencic) [reply test/lib/smartSurveyAPICall.test.js]: This issue is very similar to having proxy-require to mock smartSurveyClientAPI. There are some workarounds, but they most require to re-arrange your code (and your tests). In this particular case you could return a promise and let the user decide how they want to log the object/error.

Also, `parseResponse` is supposed to be a function, but you have a string?

File: test/lib/smartSurveyAPICall.test.js (hunk 12 above)
Comment (chrisns) [reply test/lib/smartSurveyAPICall.test.js]: @danielepolencic 
the point of the test is to assert that calling func a(1,2) results in a call to func b({apiToken:1, apiTokenSecret:2}) so I'd say its a valid test because he's testing his code, not the library

by keeping the test he'll know that if a change does happen in the library he'll know that his code works, but integration tests will pick up the fault with the library

File: lib/smartSurveyAPICall.js (hunk 8 above)
Comment (chrisns) [reply lib/smartSurveyAPICall.js]: if you do need those methods, please make a PR to `smartSurveyClient` :)

File: test/lib/smartSurveyAPICall.test.js (hunk 11 above)
Comment (chrisns) [reply test/lib/smartSurveyAPICall.test.js]: `function () {}` is still ECMA 6 syntax, it's just not an arrow function `() => {}` which as you say won't have a `this` context. and cannot be used for describing constructors which is what he's doing here.
https://www.sitepoint.com/es6-arrow-functions-new-fat-concise-syntax-javascript/
covers it reasonably well

File: lib/smartSurveyAPICall.js (hunk 8 above)
Comment (sulthan-ahmed) [reply lib/smartSurveyAPICall.js]: fair point @chrisns

File: package.json (hunk 9 above)
Comment (sulthan-ahmed) [reply package.json]: DANIELE where you have been, haven't had you review my PRs in aaages.  `mocha --recursive` worked like a charm

So I was taught that my test folder structure should follow the same structure as my app, that's why I have a lib folder.  The idea was so that it's easy to find the related test.  Is that not necessarily true? Have you got a different thought?

File: app.js [hunk 14]
```diff
@@ -21,6 +23,10 @@ app.get('/', function(req, res, next){
   }))
 })
 
+app.get('/responses', function(req, res, next){
+  res.json(smartSurveyAPICall.getData(config.apiToken, config.apiTokenSecret, config.surveyID, smartSurveyAPICall.parseResponse))
```
Comment (sulthan-ahmed) [inline app.js]: Separate PR questions: Do I need to just pass this to locals and then I can render it?

File: test/lib/smartSurveyAPICall.test.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    describe('parseResponse', function() {
+      it('should return an error message if there is an error', function() {
+        parseResponse('error', null).should.equal('error')
+      })
+
+      it('should return the result if there is no error', function() {
+        const result = parseResponse(null, 'good response');
+        const expect = JSON.stringify('good response', null, 2);
+        result.should.deep.equal(expect);
+      })
```
Comment (danielepolencic) [inline test/lib/smartSurveyAPICall.test.js]: I'm fine with not having semicolons as long as:

- you're consistent
- you know how ASI works and all its edge cases

File: lib/smartSurveyAPICall.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  getResponses: (baseUrl, surveyID, endPoint, token, tokenSecret) => {
+    querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' });
+    const apiUrlPart = querystring.stringify({'api_token':token, 'api_token_secret': tokenSecret});
+    axios.get(`${baseUrl}/${surveyID}/${endPoint}?${apiUrlPart}`)
+      .then(function (response) {
+        console.log(response);
+      })
+      .catch(function (error) {
+        console.log(error);
+      });
```
Comment (danielepolencic) [inline lib/smartSurveyAPICall.js]: This is probably very minor, buy you could use a lambda instead of an anonymous function `function () {}` => `() => {}`

File: lib/smartSurveyAPICall.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const axios = require('axios');
+const querystring = require('querystring');
+const SmartSurveyClient = require('smartsurvey-client')
+
+module.exports = {
+  getResponses: (baseUrl, surveyID, endPoint, token, tokenSecret) => {
+    querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' });
+    const apiUrlPart = querystring.stringify({'api_token':token, 'api_token_secret': tokenSecret});
```
Comment (danielepolencic) [inline lib/smartSurveyAPICall.js]: inconsistent spacing: `'api_token':token` should be `'api_token': token`.

Please consider using ESLint with a preset such as https://www.npmjs.com/package/eslint-config-airbnb-base to enforce consistent styling.

File: test/lib/smartSurveyAPICall.test.js [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+      smartSurveyAxiosAPI.getResponses(baseUrl, surveyID, endPoint, token, tokenSecret);
+      stubAxios.get.should.have.been.calledWithExactly('a/b/c?api_token=d&api_token_secret=e');
+      stub.restore();
+    });
+  })
+    describe('smartSurveyClientAPI', function() {
+      var that;
+      beforeEach( function() {
+        that = {
```
Comment (danielepolencic) [inline test/lib/smartSurveyAPICall.test.js]: It's really hard to grasp what `that` is referring to by just looking at its name. Please consider using a different name. What is `that` referring to?

**Review by danielepolencic — APPROVED** [review]

**Comment by wilko55** [comment]
Is it worth including a mock route that returns some json so you can predictably get a response? Or use something like https://github.com/typicode/json-server. Could be useful if the next stage is to manipulate the returned json

File: lib/smartSurveyAPICall.js [hunk 19]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+const axios = require('axios');
+const querystring = require('querystring');
+const SmartSurveyClient = require('smartsurvey-client')
```
Comment (wilko55) [inline lib/smartSurveyAPICall.js]: Missing semi-colon :)

File: lib/smartSurveyAPICall.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+        console.log(response);
+      })
+      .catch(function (error) {
+        console.log(error);
+      });
+  },
+
+  createClient: function (token, tokenSecret) {
+    this.client = new SmartSurveyClient({ apiToken: token, apiTokenSecret: tokenSecret})
+  },
```
Comment (wilko55) [inline lib/smartSurveyAPICall.js]: Inconsistent spacing here. You could go for the 
```
module.exports {
  getResponses: getResponses,
  createClient: createClient
  ...
}
```
approach which might tidy it up a little

File: package.json [hunk 21]
```diff
@@ -5,7 +5,7 @@
   "main": "app.js",
   "scripts": {
     "start": "node .",
-    "test": "NODE_ENV=test mocha",
+    "test": "NODE_ENV=test mocha --recursive",
```
Comment (wilko55) [inline package.json:8]: Do you have mocha installed globally? I had to add it to the package.json - `npm install -S -D mocha` to be able to run `npm test` on my local machine

File: lib/smartSurveyAPICall.js (hunk 19 above)
Comment (daniel-ac-martin) [reply lib/smartSurveyAPICall.js]: @sulthan-ahmed: Perhaps it's worth getting a linter in there. Are the HOF guys still using eslint?

File: test/lib/smartSurveyAPICall.test.js (hunk 18 above)
Comment (daniel-ac-martin) [reply test/lib/smartSurveyAPICall.test.js]: I'd also avoid 'that' as it is usually used for another purpose. (Creating a closure for a 'this' that would overwise get clobbered.)

File: lib/smartSurveyAPICall.js (hunk 16 above)
Comment (chrisns) [reply lib/smartSurveyAPICall.js]: in this particular scenario you could actually do `.catch(console.log)` or `console.error`

File: lib/smartSurveyAPICall.js (hunk 19 above)
Comment (chrisns) [reply lib/smartSurveyAPICall.js]: eslint ftw 👍

File: package.json (hunk 21 above)
Comment (chrisns) [reply package.json:8]: @wilko55 npm will know the path to local mocha, doesn't need to be globally installed

File: app.js (hunk 14 above)
Comment (chrisns) [reply app.js]: what does that mean?

**Comment by chrisns** [comment]
I think before merging this, you should run a linter, so all the comments around that, and the next PR we review won't be mostly linting changes.
`eslint --fix` is your friend 😄

File: package.json (hunk 21 above)
Comment (wilko55) [reply package.json:8]: :+1: this occurred around the same time my nvm and other path variables got corrupted

File: lib/smartSurveyAPICall.js (hunk 19 above)
Comment (sulthan-ahmed) [reply lib/smartSurveyAPICall.js]: Thanks guys, I've learnt my lesson.  Yes we are using eslint.

File: test/lib/smartSurveyAPICall.test.js (hunk 18 above)
Comment (sulthan-ahmed) [reply test/lib/smartSurveyAPICall.test.js]: @danielepolencic That is referring to the context so I'll change the name to that.

@daniel-ac-martin what is that usually used for?

File: app.js (hunk 14 above)
Comment (sulthan-ahmed) [reply app.js]: What I mean it's not for this PR but if I want to render the stuff from the JSON, is it simply just render

File: test/lib/smart-survey-api.test.js [hunk 22]
```diff
@@ -1,36 +1,39 @@
 'use strict';
 
-const sinon = require('sinon');  
+const sinon = require('sinon');
 const chai = require('chai').use(require('sinon-chai'));
+/* eslint-disable no-unused-vars*/
 const should = chai.should();
```
Comment (chrisns) [inline test/lib/smart-survey-api.test.js]: the linter is right, you have an unused var
you can just do `chai.should();`

File: lib/smart-survey-api.js [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+const axios = require('axios');
+const querystring = require('querystring');
+const SmartSurveyClient = require('smartsurvey-client');
+
+module.exports = {
+  getResponses: (baseUrl, surveyID, endPoint, token, tokenSecret) => {
+    querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' });
+    const apiUrlPart = querystring.stringify({'api_token': token, 'api_token_secret': tokenSecret});
+    axios.get(`${baseUrl}/${surveyID}/${endPoint}?${apiUrlPart}`)
```
Comment (chrisns) [inline lib/smart-survey-api.js]: have you done enough exploration now to use `axios` OR `SmartSurveyClient`?

File: lib/smart-survey-api.js [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+  getData: function getDataSmartSurveyClient(token, tokenSecret, surveyID, callback) {
+    this.createClient(token, tokenSecret);
+    this.client.getResponses(surveyID, { page: 1, pageSize: 25, includeLabels: false}, callback);
+  },
+  parseResponse: (err, result) => {
+    if (err) {
+      return err;
+    }
+
+      const prettyResult = JSON.stringify(result, null, 2);
```
Comment (chrisns) [inline lib/smart-survey-api.js]: indentation?

File: test/lib/.eslintrc [hunk 25]
```diff
@@ -0,0 +1,6 @@
+extends:
```
Comment (chrisns) [inline test/lib/.eslintrc]: do you really need to put these at each directory level or is the one in `test/` enough?

File: test/lib/smart-survey-api.test.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+// Proxyquire syntax: first argument: point to your module (which holds your dependency)
+// 2nd argument: An object with the key: module name to proxy & the value the empty object above
+// So this means smartSurveyAxiosAPI will now point to this empty object, i.e. a stub
+const smartSurveyAxiosAPI = proxyquire('../../lib/smart-survey-api', {
+  'axios': stubAxios,
+});
+
+// create an empty function using not using arrow functions because using context
+var stubSmartSurvey = function empty() {
```
Comment (chrisns) [inline test/lib/smart-survey-api.test.js]: why are you naming the function different to what it's assigned to?
this will be executable with both `stubSmartSurvey()` and `empty()` which is probably not what you intended

File: test/lib/smart-survey-api.test.js [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+      // this works as well, stubAxios.get = sinon.stub().resolves('test')
+      // however, this is done so that the stub can be restored for later
+      let stub = sinon.stub(stubAxios, 'get').resolves('test');
+
+      smartSurveyAxiosAPI.getResponses(baseUrl, surveyID, endPoint, token, tokenSecret);
+      stubAxios.get.should.have.been.calledWithExactly('a/b/c?api_token=d&api_token_secret=e');
+      stub.restore();
+    });
+  });
+    describe('smartSurveyClientAPI', function() {
```
Comment (chrisns) [inline test/lib/smart-survey-api.test.js]: indentation?

File: test/lib/smart-survey-api.test.js [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+          smartSurveyClientAPI = proxyquire('../../lib/smart-survey-api', {
+            'smartsurvey-client': stubSmartSurvey
+          });
+
+          smartSurveyClientAPI.createClient.call(context, '1', '2');
+          stubSmartSurvey.should.have.been.called;
+      });
+  });
+
+      it('should call the callback with an error response when we receive an error from the api', function() {
```
Comment (chrisns) [inline test/lib/smart-survey-api.test.js]: indentation?

**Review by chrisns — CHANGES_REQUESTED** [review]

File: lib/smart-survey-api.js (hunk 23 above)
Comment (sulthan-ahmed) [reply lib/smart-survey-api.js]: Is there any harm keeping both? There's an api call that isn't covered for SmartSurveyClient where I need to use axios.  

I know you suggested putting in a PR for SmartSurveyClient.  That I'll have to put in as a later stage, don't you consider this PR out of scope for that?

Comment (chrisns) [reply lib/smart-survey-api.js]: what is the api call that is not covered by SmartSurveyClient?

File: test/lib/.eslintrc (hunk 25 above)
Comment (sulthan-ahmed) [reply test/lib/.eslintrc]: ahh good spot, I don't know how that happened

File: test/lib/smart-survey-api.test.js (hunk 26 above)
Comment (sulthan-ahmed) [reply test/lib/smart-survey-api.test.js]: yeah definitely wasn't my intention, the linter was complaining about anonymous functions so I was just giving it a name

Comment (chrisns) [reply test/lib/smart-survey-api.test.js]: http://eslint.org/docs/rules/func-names
the `as-needed` seems like a pragmatic option for you, looks like you've maybe got an 'always' set?

File: lib/smart-survey-api.js (hunk 23 above)
Comment (sulthan-ahmed) [reply lib/smart-survey-api.js]: anything to do with tracking links https://docs.smartsurvey.io/v1/reference#links

File: test/lib/smart-survey-api.test.js (hunk 26 above)
Comment (sulthan-ahmed) [reply test/lib/smart-survey-api.test.js]: Surprisingly, I tried that and I got more linting issues requesting not to have unnamed functions in my describe blocks and it blocks

File: .gitignore [hunk 29]
```diff
@@ -1,3 +1,4 @@
 node_modules
 public
 npm-debug.log
+package-lock.json
```
Comment (chrisns) [inline .gitignore]: why do you want to ignore this?

**Review by chrisns — CHANGES_REQUESTED** [review]

File: lib/smart-survey-api.js (hunk 23 above)
Comment (chrisns) [reply lib/smart-survey-api.js]: if you need to use that, then you can easily either augment/extend or fork+PR the request for that and still benefit from the smartsurveyclient's pagination, auth and other abstractions, rather than having to do that yourself

Comment (sulthan-ahmed) [reply lib/smart-survey-api.js]: thanks @chrisns 
I'm not sure how to augment/extend it, is there any harm accepting this PR as it is and then I can do a separate PR for what you are suggesting? 
I'm just wondering if this PR is just getting bigger and bigger than what it needs to be?

File: .gitignore (hunk 29 above)
Comment (sulthan-ahmed) [reply .gitignore]: A lot of dependency issues that we've experienced in a ton of apps for service ops.  e.g.
https://github.com/npm/npm/issues/16866

I can remove this because I don't think this change needs to be discussed as part of this PR

File: lib/smart-survey-api.js (hunk 23 above)
Comment (chrisns) [reply lib/smart-survey-api.js]: you're not doing anything with `links` yet. so arguably you could make the PR **smaller** by removing the stuff you don't need *yet*

Comment (sulthan-ahmed) [reply lib/smart-survey-api.js]: I'm happy to remove it.  

Is there anyway, I could keep the code for future reference?

Comment (chrisns) [reply lib/smart-survey-api.js]: its in the git history :)

Comment (sulthan-ahmed) [reply lib/smart-survey-api.js]: @easternbloc has a good idea, I'll add a new commit and delete it

**Comment by sulthan-ahmed** [comment]
@chrisns has sadly left us 😢 However, I've addressed the issues he has raised. @danielepolencic has also approved the PR so I will now merge this PR 🎆
